src/NHibernate/Proxy/NHibernateProxyBuilder.cs

```diff
@@ -0,0 +1,362 @@
+
+			var interfaces = new HashSet<System.Type>
+			{
+				// Add the ISerializable interface so that it can be implemented
+				typeof(ISerializable)
+			};
+			interfaces.UnionWith(baseInterfaces);
+			interfaces.UnionWith(baseInterfaces.SelectMany(i => i.GetInterfaces()));
+			interfaces.UnionWith(baseType.GetInterfaces());
+
+			// Use the object as the base type 
+			// since we're not inheriting from any class type
+			var parentType = baseType;
+			if (baseType.IsInterface)
+			{
+				parentType = typeof(object);
+				interfaces.Add(baseType);
+			}
+
+			var typeBuilder = moduleBuilder.DefineType(typeName, typeAttributes, parentType, interfaces.ToArray());
+
+			var lazyInitializerField = typeBuilder.DefineField("__lazyInitializer", LazyInitializerType, FieldAttributes.Private);
+			var proxyInfoField = typeBuilder.DefineField("__proxyInfo", typeof(NHibernateProxyFactoryInfo), FieldAttributes.Private);
+
+			ImplementConstructor(typeBuilder, parentType, lazyInitializerField, proxyInfoField);
+
+			// Provide a custom implementation of ISerializable
+			// instead of redirecting it back to the interceptor
+			foreach (var method in ProxyFactory.GetProxiableMethods(baseType, interfaces.Except(new[] {typeof(ISerializable)})))
+			{
+				CreateProxiedMethod(typeBuilder, method, lazyInitializerField);
+			}
+
+			// Make the proxy serializable
+			var serializableConstructor = typeof(SerializableAttribute).GetConstructor(System.Type.EmptyTypes);
+			var customAttributeBuilder = new CustomAttributeBuilder(serializableConstructor, new object[0]);
+			typeBuilder.SetCustomAttribute(customAttributeBuilder);
+
+			ImplementDeserializationConstructor(typeBuilder);
+			ImplementGetObjectData(typeBuilder, proxyInfoField, lazyInitializerField);
+
+			var proxyType = typeBuilder.CreateTypeInfo();
+
+			ProxyAssemblyBuilder.Save(assemblyBuilder);
+
+			return proxyType;
+		}
+
+		void CreateProxiedMethod(TypeBuilder typeBuilder, MethodInfo method, FieldInfo lazyInitializerField)
+		{
+			if (method == NHibernateProxyTypeLazyInitializerProperty.GetMethod)
+			{
+				ImplementGetLazyInitializer(typeBuilder, method, lazyInitializerField);
+			}
+			else if (method == _getIdentifierMethod)
+			{
+				ImplementGetIdentifier(typeBuilder, method, lazyInitializerField);
+			}
+			else if (method == _setIdentifierMethod)
+			{
+				ImplementSetIdentifier(typeBuilder, method, lazyInitializerField);
+			}
+			else if (!_overridesEquals && method.Name == "Equals" && method.GetBaseDefinition() == typeof(object).GetMethod("Equals", new[] {typeof(object)}))
+			{
+				//skip
+			}
+			else if (!_overridesEquals && method.Name == "GetHashCode" && method.GetBaseDefinition() == typeof(object).GetMethod("GetHashCode"))
+			{
+				//skip
+			}
+			else if (_componentIdType != null && _componentIdType.IsMethodOf(method))
+			{
+				ImplementCallMethodOnEmbeddedComponentId(typeBuilder, method, lazyInitializerField);
```

Comment:
Not sure what is that, or how it's supposed to be used. I've copied this from DefaultProxyFactory. Do you think that this path needs to be removed?

Reply:
With what I have seen on Hibernate side, it now looks to me as something partially ported, causing it to very likely never have any usage. Fully ascertaining this would require a bit more digging on Hibernate side to see in which cases it serves.
I am always for removing useless code, but since it does not cause any harm, it also not really worth it to dig more for checking it can really be removed without any risk.

So if you are not sure about it either, it is going to stay for now.